### PR TITLE
chore(mappings): POC for help and :help on_attach keymaps

### PR DESCRIPTION
**Pull request by alex-courtis**
1. generate help view by mapping keymap functions back to `DEFAULT_KEYMAPS`
2. generate :help via update script
a. user readable table `/tmp/DEFAULT_KEYMAPS.help`
b. on_attach function `/tmp/DEFAULT_KEYMAPS.on_attach.lua`

I did mess around with `callback` as a string to loadstring. It works, however I lost the ability to map back, as the function from loadstring is a new wrapper function.

We don't need the complication of loadstring in code, so text processing in the update script is enough. If necessary, we can apply the same processing to generate lua to perform the legacy conversion.

### CODE REVIEW

File: lua/nvim-tree/renderer/help.lua [hunk 1]
```diff
@@ -1,27 +1,62 @@
...
-    local name
-    if cb and cb:sub(1, 35) == require("nvim-tree.config").nvim_tree_callback("test"):sub(1, 35) then
-      name = cb:match "'[^']+'[^']*$"
-      name = name:match "'[^']+'"
-    elseif b.action then
-      name = b.action
+  for _, bkm in ipairs(buf_keymaps) do
+    local default_keymap = nil
+    for _, dkm in ipairs(require("nvim-tree.keymap").DEFAULT_KEYMAPS) do
+      if bkm.callback == dkm.callback then
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/help.lua:28]: Very happy with this. It successfully maps user's on_attach mappings as well as defaults.

`inject_node` was not a problem as it returns an anonymous function that mapped well.

File: lua/nvim-tree/renderer/help.lua [hunk 2]
```diff
@@ -1,27 +1,62 @@
...
+  --   elseif b.action then
+  --     name = b.action
+  --   else
+  --     name = (b.name ~= nil) and b.name or cb
+  --     name = '"' .. name .. '"'
+  --   end
+  --   table.insert(processed, { key, name, true })
+  -- end
+
   table.sort(processed, function(a, b)
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/help.lua:60]: This sort order looks a bit strange now; maybe we switch to ordering by key.

**Review by kyazdani42 — COMMENTED** [review]
I think we should probably adapt this to attach short descriptions on the keymap directly so we don't need to do some custom logic :)
see `:help vim.api.nvim_set_keymap`, these applies to vim.keymap.set as well.
This way we could just set the same description from the old mapping table, and the new mapping table, so we can keep retro compatibility while moving to the new ones.

File: lua/nvim-tree/renderer/help.lua (hunk 2 above)
Comment (kyazdani42) [reply lua/nvim-tree/renderer/help.lua:60]: indeed, simplify this please !

File: scripts/generate_default_keymaps.lua [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    io.write(string.format(fmt, m.key, m.short, m.long))
+  end
+end
+io.write "\n"
+io.close(file)
+
+-- lua on_attach
+file = io.open("/tmp/DEFAULT_KEYMAPS.on_attach.lua", "w")
+io.output(file)
+io.write "local function on_attach(bufnr, mode, opts)\n"
```
Comment (kyazdani42) [inline scripts/generate_default_keymaps.lua:45]: why are mode and opts passed to on_attach ? They should be unique for each mapping i believe.

**Comment by alex-courtis** [comment]
> I think we should probably adapt this to attach short descriptions on the keymap directly so we don't need to do some custom logic :) see `:help vim.api.nvim_set_keymap`, these applies to vim.keymap.set as well. This way we could just set the same description from the old mapping table, and the new mapping table, so we can keep retro compatibility while moving to the new ones.

Oh wow... that makes things easy.

File: scripts/generate_default_keymaps.lua (hunk 3 above)
Comment (alex-courtis) [reply scripts/generate_default_keymaps.lua:45]: I'm not sure that the user should care about mode and opts... we should pass in sensible defaults and they can change if they want to.

RE mode: I don't think anything but "n" makes sense.

File: lua/nvim-tree/renderer/help.lua (hunk 2 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/help.lua]: It's now ordering something like :help index

File: scripts/generate_default_keymaps.lua (hunk 3 above)
Comment (alex-courtis) [reply scripts/generate_default_keymaps.lua]: Updated to include desc, removed mode and opts.

**Comment by alex-courtis** [comment]
> This way we could just set the same description from the old mapping table, and the new mapping table, so we can keep retro compatibility while moving to the new ones.

Working nicely. desc is set to old mapping action and new keymap desc.short.

File: scripts/generate_default_keymaps.lua (hunk 3 above)
Comment (kyazdani42) [reply scripts/generate_default_keymaps.lua]: yes 'n' by default, but the old bindings could have mode associated with them (if the user wishes to do stuff with the visual selection).

**Review by kyazdani42 — APPROVED** [review]
this is much better. Since i won't be too much active until end of august, i'll leave you deciding if you want to merge :)

File: lua/nvim-tree/renderer/help.lua [hunk 4]
```diff
@@ -1,48 +1,96 @@
...
+local function shorten_lhs(lhs)
+  return lhs
+    :gsub("LeftMouse>$", "LM>")
+    :gsub("RightMouse>$", "RM>")
+    :gsub("MiddleMouse>$", "MM>")
+    :gsub("ScrollWheelDown>$", "SD>")
+    :gsub("ScrollWheelUp>$", "SU>")
+    :gsub("ScrollWheelLeft>$", "SL>")
+    :gsub("ScrollWheelRight>$", "SR>")
+end
```
[the remark this answers is not in the document]
Comment (alex-courtis) [reply lua/nvim-tree/renderer/help.lua:29]: I can revert. Now that mouse is at the end, the indenting won't look so strange.

**Comment by alex-courtis** [comment]
> this is much better. Since i won't be too much active until end of august, i'll leave you deciding if you want to merge :)

I will not merge, it's not urgent. I might collect all our work in a single branch that we can review as one when you return.

**Comment by alex-courtis** [comment]
See #1579

**Pull request closed without merging**